File: algorithms/BinaryTree/PathSum.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 *      Runtime: 286 ms
 * }
 */
public class Solution {
    public boolean hasPathSum(TreeNode root, int sum) {
        Stack<TreeNode> st = new Stack<TreeNode>();
        Stack<Boolean> left = new Stack<Boolean>();
        Stack<Boolean> right = new Stack<Boolean>();
        TreeNode curr = root;
        int res = 0;
        while(!st.empty() || curr != null)
        {
        	while(curr != null)
        	{
        	    res += curr.val;
        	    st.push(curr);
        	    left.push(true);
        	    right.push(false);
        	    
        	    curr = curr.left;
        	}
        	TreeNode leaf = st.peek();
        	if(res == sum && leaf.left == null && leaf.right == null)
        	{
        		    return true;
        	}
        	else
        	{
                curr = st.peek();  
                
                if(curr.right != null && right.peek() == false)
                {  
                	right.pop();
                	right.push(true);
                    curr = curr.right;
                    //st.push(curr);
                    //res += curr.val;
                    //left.push(true);
                    //right.push(false);
                }  
                else
                {  
                	res -= st.pop().val;
                	left.pop();
                	right.pop();
                	curr = null;
                }
        	}
        }
        return false;
    }
}

File: algorithms/BinaryTree/PathSum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 * Runtime: 12 ms
 */
class Solution {
public:
	int res;
	bool flag;
    bool hasPathSum(TreeNode* root, int sum) {
		if(root == NULL)
			return false;
		flag = false;
        res = sum;
		currSum(root,0);
		return flag;
    }
	void currSum(TreeNode* root, int curr)
	{
		if(flag)
			return;
		if(root->left==NULL && root->right ==NULL)
		{
			curr += root->val;
			if(curr == res)
				flag = true;
			return ;
		}
		if(root->left != NULL)
		{
			currSum(root->left,curr+root->val);
		}
		if(root->right != NULL)
		{
			currSum(root->right,curr+root->val);
		}
	}
};